### ISSUE TITLE
Moving icons corrupts connections

### ISSUE THREAD

**Issue by perost** [OpenModelica/OpenModelica#11919]
### Description

Moving icons around in a diagram and then deleting connections can cause zombie connections, invalid connections, and even crashes.

### Steps to Reproduce

Create a new model with some components and connect them together, for example:
![img1](https://github.com/OpenModelica/OpenModelica/assets/4406023/1c73716d-ac1a-4c9c-aa83-18a0d7ba8251)
Move one of the icons around by just dragging it with the mouse. Moving the icon multiple times might be required to trigger the bug, though sometimes it happens by just moving it once.
![img2](https://github.com/OpenModelica/OpenModelica/assets/4406023/691c4659-ee72-4bfa-9da1-ca5faf451d53)
Select the connection and delete it. This can cause the connection to still remain as a "zombie" that can't be selected and which disappears when the icon is moved again, or it can cause OMEdit to crash, or it can change the connection like this:
![img3](https://github.com/OpenModelica/OpenModelica/assets/4406023/c8be052f-cc8c-45dc-b388-bd271602e08d)

### Expected Behavior

Deleting connections to icons that have been moved around should work as normal without crashing or other issues.

### Version and OS

 - OpenModelica Version: master
 - OS: Linux

**Comment by perost** [comment]
Valgrind says:
```
==2438961== Invalid read of size 8
==2438961==    at 0xA79BAA8: QGraphicsItem::sceneBoundingRect() const (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0x5433BB: LineAnnotation::adjustPointsForDrawing() const (LineAnnotation.cpp:1480)
==2438961==    by 0x53CF5C: LineAnnotation::getShape() const (LineAnnotation.cpp:664)
==2438961==    by 0x53D3F6: LineAnnotation::shape() const (LineAnnotation.cpp:705)
==2438961==    by 0x53D34B: LineAnnotation::boundingRect() const (LineAnnotation.cpp:700)
==2438961==    by 0xA7A27E6: QGraphicsItemPrivate::effectiveBoundingRect(QGraphicsItem*) const (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0xA7CFE2B: ??? (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0xA7F2FF3: QGraphicsView::paintEvent(QPaintEvent*) (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0xA4D803F: QWidget::event(QEvent*) (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0xA5869D2: QFrame::event(QEvent*) (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==    by 0xB7795C1: QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) (in /usr/lib/libQt5Core.so.5.15.11)
==2438961==    by 0xA4A18EE: QApplicationPrivate::notify_helper(QObject*, QEvent*) (in /usr/lib/libQt5Widgets.so.5.15.11)
==2438961==  Address 0x9a9b9838 is 24 bytes inside a block of size 1,248 free'd
==2438961==    at 0x484561D: operator delete(void*, unsigned long) (vg_replace_malloc.c:1101)
==2438961==    by 0x311398: Element::~Element() (Element.h:187)
==2438961==    by 0x50D5BC: Element::removeChildrenNew() (Element.cpp:1729)
==2438961==    by 0x50D500: Element::removeChildrenNew() (Element.cpp:1723)
==2438961==    by 0x50D500: Element::removeChildrenNew() (Element.cpp:1723)
==2438961==    by 0x50D780: Element::reDrawElementNew() (Element.cpp:1740)
==2438961==    by 0x3D81DC: GraphicsView::drawElements(ModelInstance::Model*, bool, ModelInfo const&) (ModelWidgetContainer.cpp:446)
==2438961==    by 0x40272F: ModelWidget::drawModelIconDiagram(ModelInstance::Model*, bool, ModelInfo const&) (ModelWidgetContainer.cpp:5995)
==2438961==    by 0x40252A: ModelWidget::drawModel(ModelInfo const&) (ModelWidgetContainer.cpp:5978)
==2438961==    by 0x402A3E: ModelWidget::loadModelInstance(bool, ModelInfo const&) (ModelWidgetContainer.cpp:6025)
==2438961==    by 0x409D47: ModelWidget::reDrawModelWidget(ModelInfo const&) (ModelWidgetContainer.cpp:6777)
==2438961==    by 0x81CE11: OMCUndoCommand::redoInternal() (Commands.cpp:1678)
==2438961==  Block was alloc'd at
==2438961==    at 0x4841FA3: operator new(unsigned long) (vg_replace_malloc.c:483)
==2438961==    by 0x50C542: Element::createClassElements() (Element.cpp:1579)
==2438961==    by 0x50C3D5: Element::createClassElements() (Element.cpp:1570)
==2438961==    by 0x50C3D5: Element::createClassElements() (Element.cpp:1570)
==2438961==    by 0x5128BC: Element::drawModelicaElement() (Element.cpp:2413)
==2438961==    by 0x512325: Element::drawElement() (Element.cpp:2362)
==2438961==    by 0x50D8BF: Element::reDrawElementNew() (Element.cpp:1749)
==2438961==    by 0x3D81DC: GraphicsView::drawElements(ModelInstance::Model*, bool, ModelInfo const&) (ModelWidgetContainer.cpp:446)
==2438961==    by 0x40272F: ModelWidget::drawModelIconDiagram(ModelInstance::Model*, bool, ModelInfo const&) (ModelWidgetContainer.cpp:5995)
==2438961==    by 0x40252A: ModelWidget::drawModel(ModelInfo const&) (ModelWidgetContainer.cpp:5978)
==2438961==    by 0x402A3E: ModelWidget::loadModelInstance(bool, ModelInfo const&) (ModelWidgetContainer.cpp:6025)
==2438961==    by 0x409D47: ModelWidget::reDrawModelWidget(ModelInfo const&) (ModelWidgetContainer.cpp:6777)
```

**Comment by casella** [comment]
We just got a crash report with a similar set up, using 1.21.1. I invited the reporter to open a ticket about it. This is the outcome of the crash reporting tool:

https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-omeditcommunication.log
https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-omeditcommands.mos
https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-openmodelica.stacktrace.OMEdit

I guess it could be related

**Comment by perost** [comment]
> We just got a crash report with a similar set up, using 1.21.1. I invited the reporter to open a ticket about it. This is the outcome of the crash reporting tool:
> 
> https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-omeditcommunication.log https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-omeditcommands.mos https://dev.openmodelica.org/omeditcrashreports/tmp/2024-01-31_21:18:41-openmodelica.stacktrace.OMEdit
> 
> I guess it could be related

Yes, we've gotten a lot of reports about it, but that was the first report with some actual description of what the user was doing when encountering the issue. I then stumbled on this way to reproduce it while I was playing around with it. I've already informed the user that I opened this issue.

**Comment by adeas31** [comment]
I will take a look at it asap.

**Comment by adeas31** [comment]
@perost PR #11936 is up for testing.

It was very difficult to debug this issue. An extra paint event was triggered for the deleted connection. I hope it is fixed now with my changes. I tested several times without any crash.

**Comment by perost** [comment]
> @perost PR #11936 is up for testing.
> 
> It was very difficult to debug this issue. An extra paint event was triggered for the deleted connection. I hope it is fixed now with my changes. I tested several times without any crash.

Great! I just tested your fix and it seems to be working fine now.

**Comment by adeas31** [comment]
Keep it open until it is ported to 1.22

**Comment by ShuffleWire** [comment]
Hello ! 
`We just got a crash report with a similar set up, using 1.21.1. I invited the reporter to open a ticket about it. This is the outcome of the crash reporting tool:`
I'm the reporter of the issue, and you are on fire xD Didn't expect such a quick reply ! 
I do think that opening an issue is not relevant anymore, considering that the issue seems to be already solved.
Thank again, it actually a very pleasant thing to see such an activity.
